mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala

```diff
@@ -562,4 +562,20 @@ trait HasBlockSize extends Params {
   /** @group expertGetParam */
   final def getBlockSize: Int = $(blockSize)
 }
+
+/**
+ * Trait for shared param k. This trait may be changed or
+ * removed between minor versions.
+ */
+trait HasK extends Params {
+
+  /**
+   * Param for The number of clusters to create. Must be (&gt; 1). Note that it is possible for fewer than k clusters to be returned.
+   * @group param
+   */
+  final val k: IntParam = new IntParam(this, "k", "The number of clusters to create. Must be (> 1). Note that it is possible for fewer than k clusters to be returned", ParamValidators.gt(1))
```

Comment:
Hm, I guess we lose the `@Since` annotations here. The latest one was 2.4.0. I guess we could say these methods are since 2.4.0, but the trait is since 3.1.0, which might be confusing. I don't think it's a big deal but is a minor downside to this refactoring.

Reply:
Yes, we lost the ```@Since``` annotations for ```k``` and ```getK```. We still have since 2.4.0 for ```setK```, though. 
For all the shared params, we don't have ```@Since``` annotations for the param and getXXX, only have ```@Since``` annotations for setXXX.